docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,518 @@
+
+```yaml
+kind: HostnameGenerator
+metadata:
+  name: k8s-zone-hostnames
+spec:
+  targetRef:
+    kind: MeshExternalService
+    tags:
+      my.label.io/service: my-service 
+  template: {{ name }}.svc.meshext.local
+```
+
+For example, given this `MeshExternalService`
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: mydomain
+  namespace: kuma-system
+  labels:
+    my.label.io/service: my-service 
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+    kuma.io/origin: zone
+spec:
+  destinations:
+  - address: 192.168.0.1
+    port: 9090 
+...
+```
+
+We would generate such hostname `mydomain.svc.meshext.local`
+
+Possible template functions/keys are:
+
+`{{ name }}` - name of the `MeshExternalService`
+`{{ label "x" }}` - value of label x.
+
+If the template cannot be resolved (label is missing), the hostname won't be generated with a NotAvailable status and an appropriate reason.
+
+#### Positive Consequences
+
+* Simpler model, but not as clear
+* One domain per external service 
+
+#### Negative Consequences
+
+* Using HostnameGenerator might cause unexpected domains to be created, when match is too wide
+
+#### Creating MeshExternalService and using HostnameGenerator for InternalVIP
+
+This option is similar to the chosen one, but instead of explicitly providing a domain in the policy, users would need to create a `HostnameGenerator` to generate a domain.
+
+The API model does not include a `value` field for the `InternalVIP` type. The domain name is generated by `HostnameGenerator` and later provided under a status section.
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: InternalVIP # Kuma will generate a domain based on HostnameGenerator
+    port: 80
+    protocol: http
+...
+status:
+  addresses:
+  - hostname: httpbin.svc.meshext.local
+    status: Available # | NotAvailable
+    origin: 
+      kind: HostGenerator
+      name: "k8s-zone-generator"
+    reason: "not available because of the clash with ..."
+  vips:
+  - ip: <kuma VIP>
+    type: Kuma | # External
+```
+
+#### Positive Consequences
+
+* Separation between hostname generation and possible better security (but not a huge problem when only MeshOperator is allowed to create the policy)
+
+#### Negative Consequences
+
+* Less explicit
+* Match section looks a bit awkward with a hostname generator
```

Comment:
I think our core problem is that you can have multiple Matches. If you have a single match it's not awkward IMO:

```yaml
kind: MeshExternalService
metadata:
  name: example
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
    kuma.io/zone: east-1
spec:
  match:
    type: KumaHostname # Kuma will generate a domain based on HostnameGenerator
    port: 80
    protocol: http
  destinations:
    - address: 1.1.1.1
      port: 12345
```

Reply:
```yaml
kind: MeshPassthrough
name: my-passthrough
spec:
  targetRef:
     kind: MeshSubset
     tags:
        has-access-to-github: "true"
  spec:
    matchAppend:
    - type: Domain
      value: api.github.com
      port: 443
      protocol: tls
```

And then someone else would have:

```yaml
kind: MeshPassthrough
name: my-openai
spec:
  targetRef:
     kind: MeshSubset
     tags:
        has-access-to-openai: "true"
  spec:
    matchAppend:
    - type: Domain
      value: api.chatgpt.com
      port: 443
      protocol: tls
```